controllers/effective_tls_policies_reconciler.go

```diff
@@ -0,0 +1,394 @@
+
+func (t *EffectiveTLSPoliciesReconciler) Reconcile(ctx context.Context, events []controller.ResourceEvent, topology *machinery.Topology, _ error, s *sync.Map) error {
+	logger := controller.LoggerFromContext(ctx).WithName("EffectiveTLSPoliciesReconciler").WithName("Reconcile")
+
+	// Get affected TLS Policies
+	policies := GetTLSPoliciesByEvents(topology, events)
+
+	// Get all certs in topology for comparison with expected certs to determine orphaned certs later
+	certs := lo.FilterMap(topology.Objects().Items(), func(item machinery.Object, index int) (*certmanv1.Certificate, bool) {
+		r, ok := item.(*controller.RuntimeObject)
+		if !ok {
+			return nil, false
+		}
+		c, ok := r.Object.(*certmanv1.Certificate)
+		if !ok {
+			return nil, false
+		}
+
+		// Only want certs owned by TLSPolicies
+		if isObjectOwnedByGroupKind(c, kuadrantv1alpha1.TLSPolicyGroupKind) {
+			return c, true
+		}
+
+		return nil, false
+	})
+
+	var expectedCerts []*certmanv1.Certificate
+
+	for _, p := range policies {
+		policy := p.(*kuadrantv1alpha1.TLSPolicy)
+
+		// Get all listeners where the gateway contains this policy
+		// TODO: Update when targeting by section name is allowed, the listener will contain the policy rather than the gateway
+		listeners := lo.FilterMap(topology.Targetables().Items(), func(t machinery.Targetable, index int) (*machinery.Listener, bool) {
+			l, ok := t.(*machinery.Listener)
+			return l, ok && lo.Contains(l.Gateway.Policies(), p)
+		})
+
+		// Policy is deleted
+		if policy.DeletionTimestamp != nil {
+			logger.V(1).Info("policy is marked for deletion, nothing to do", "name", policy.Name, "namespace", policy.Namespace, "uid", policy.GetUID())
+			continue
+		}
+
+		// Policy is not valid
+		isValid, _ := IsTLSPolicyValid(ctx, s, policy)
+		if !isValid {
+			logger.V(1).Info("deleting certs for invalid policy", "name", policy.Name, "namespace", policy.Namespace, "uid", policy.GetUID())
+			if err := t.deleteCertificatesForPolicy(ctx, topology, policy); err != nil {
+				logger.Error(err, "unable to delete certs for invalid policy", "name", policy.Name, "namespace", policy.Namespace, "uid", policy.GetUID())
+			}
+			continue
+		}
+
+		// Policy is valid
+		for _, l := range listeners {
+			// Need to use Gateway as listener hosts can be merged into a singular cert if using the same cert reference
+			expectedCertificates := expectedCertificatesForGateway(ctx, l.Gateway.Gateway, policy)
+
+			for _, cert := range expectedCertificates {
+				resource := t.client.Resource(CertManagerCertificatesResource).Namespace(cert.GetNamespace())
+
+				// Check is cert already in topology
+				objs := topology.Objects().Children(l)
+				obj, ok := lo.Find(objs, func(o machinery.Object) bool {
+					return o.GroupVersionKind().GroupKind() == CertManagerCertificateKind && o.GetNamespace() == cert.GetNamespace() && o.GetName() == cert.GetName()
+				})
+
+				// Create
+				if !ok {
```

Comment:
Sure but whether to do `!ok` or `ok` first depends also on whether we `create` or `update` first, so if we want to be more positive in our check here `if ok`, this would mean `update` is first 🤔 

Not sure which action do we expect more of. Maybe @mikenairn can answer on this

Reply:
I wouldn't expect too many updates of a Certificate resource after creation. I don't have strong opinions on this, but for what its worth i did `if ok` in the dnspolicy equivalent task, and deal with updates first.

Reply:
`if ok { update }` / `if !ok { create }`. Does it matter? The number of jump instructions in the final compiled code is literally the same I think.

If `!ok` feels like being negative (it doesn't to me BTW), then perhaps we could rename the variable to `missing`. Is it better? I.e. `if missing { create }`.

IMO, if it doesn't have any obvious downside in performance, the important thing is that the code is readable. In my head (and it could be only me), I read this in terms of chronology and progressive use cases.

The first thing that happens is having nothing, then:
1. user creates the network resources
2. user creates the policies
3. the operator creates the internal resources (for the first time after seeing the policies)
4. user adjusts the policies and/or network resources
5. operator updates the internal resources accordingly.

Reply:
Thanks guys 👍  I don't think it matters in fairness unless people really feel it hinders readability. My preference is to keep it as is tbh as I also prefer dealing with the `create` first followed by `update` action. If we feel strongly about this, maybe this can be done in a separate where this can be dealt with in the entire codebase